build.assets/tooling/cmd/gen-event-reference/README.md

```diff
@@ -0,0 +1,61 @@
+# Audit event reference generator
+
+The audit event reference generator produces a list of audit events for
+inclusion in the documentation.
+
+## Background
+
+Teleport audit events have unique codes (e.g., `TDB01I`) as well as types (e.g.,
+`user.login`) that are not necessarily unique. Teleport transmits audit events
+as protocol buffer messages that Teleport audit log represents as JSON messages.
+Therefore, to represent audit events, we need to associate codes and types with
+JSON schemas.
+
+The Teleport source assigns audit event codes and types as struct values when
+initializing an audit event. Otherwise, there is nothing intrinsic to an audit
+event that associates its code, type, and schema. Finding all parts of the
+source that emit an audit event is infeasible, so the best we can do is to take
+advantage of the naming conventions we use for declarations of types, codes, and
+schemas.
+
+The reference generator assumes that event codes are constants declared in a
+single file with a declaration name in the following format:
+
+```
+CamelCaseName(Success|Failure)?Code
+```
+
+`CamelCaseName` is a prefix we expect to find across the names of event codes,
+types, and schemas.
+
+The event type that corresponds to a code, also declared as a constant in a
+single Go file, has a declaration name with the following format:
+
+```
+CamelCaseNameEvent
+```
+
+Finally, the generator expects audit event schemas to be declared as protobuf
+messages with declaration names in the following format:
+
+```
+CamelCaseName
+```
+
+Not all audit event codes, types, and schemas follow this convention, but enough
+events do follow it that we can iterate to either adjust the naming convention
```

Comment:
When I run this, it works on 95 of 247 attempts, which is a < 40% success rate.

I'm not expecting this to be able to work for anywhere near 100% of cases, but if we can't at least cover the majority are we any better off than we are today? We will still have a massive number of undocumented events.

I wonder if this is a viable approach or if there are better alternatives.

For example, `makeEvent.ts` has a strongly typed list of all events that we expect the UI to be able to render. We do occasionally forget to add events here, but this has been rare since we got rid of the separate webapps repo - I'm confident we get it right wayyy more than 40% of the time.

What if we added a new description field to the `Formatters` type in our UI code, and then wrote the reference generator in typescript instead of Go. Advantages of this approach:
- we can start with the field being optional and later make it required
- once the field is required, we get compile-time errors when the description is missing (this is something the Go-based generator here cannot enforce since it's based on conventions/comments and not leveraging the type system)